Optimize character escaping.

reitermarkus opened the pull request:
Allow optimization of panicking branch in `EscapeDebug`, see https://github.com/rust-lang/rust/pull/121805.

r? @joboet

reitermarkus committed:
Inline `EscapeDebug::size_hint`.

joboet commented:
I'm so sorry, I had another look at this and found that my analysis was incorrect: I forgot to switch the compiler...

The actual cause (fingers crossed) is in `EscapeIterInner::next` (and the same applies to its `next_back` and `as_ascii` methods). The compiler does not know about this range invariant:
https://github.com/rust-lang/rust/blob/40dcd796d094b911b8b7b55a0519fb8e3d21680f/library/core/src/escape.rs#L70
so it cannot optimize out the branch checks here:
https://github.com/rust-lang/rust/blob/40dcd796d094b911b8b7b55a0519fb8e3d21680f/library/core/src/escape.rs#L102
([Godbolt example](https://rust.godbolt.org/z/6ff7edabK))

The [`ufmt` example](https://rust.godbolt.org/z/s74qsE5Yd) iterates over an escaped `EscapeDebug` and therefore calls `EscapeIterInner::next`, leading to the panic branch.

You are welcome to fix this in this PR, but can also open another one and just assign me.

reitermarkus committed:
Avoid panicking branch in `EscapeIterInner`.

reitermarkus commented:
@joboet, I have now refactored `EscapeIterInner` to better contain the `alive` invariant and changed the index operations to use `get_unchecked`.

joboet commented on library/core/src/escape.rs:
```diff
@@ -6,56 +6,85 @@ use crate::ops::Range;
+    output[1] = ascii::Char::SmallU;
+    output[2] = ascii::Char::LeftCurlyBracket;
+
+    output[3 + u_len.saturating_sub(6)] = HEX_DIGITS[((c >> 20) & 0x0f) as usize];
+    output[3 + u_len.saturating_sub(5)] = HEX_DIGITS[((c >> 16) & 0x0f) as usize];
+    output[3 + u_len.saturating_sub(4)] = HEX_DIGITS[((c >> 12) & 0x0f) as usize];
+    output[3 + u_len.saturating_sub(3)] = HEX_DIGITS[((c >> 8) & 0x0f) as usize];
+    output[3 + u_len.saturating_sub(2)] = HEX_DIGITS[((c >> 4) & 0x0f) as usize];
+    output[3 + u_len.saturating_sub(1)] = HEX_DIGITS[((c >> 0) & 0x0f) as usize];
```
I find the `saturating_sub` dance here quite confusing. Would it be possible to go back to the previous version?

reitermarkus replied:
The previous version puts the output at the end of the buffer and returns a range, while this puts the output at the start and only needs to return the length.

reitermarkus replied:
It can be refactored to use `u_len` for the shift instead of the index, maybe that makes it a bit easier to read.

joboet replied:
I'd prefer that, it seems neater.

reitermarkus replied:
I also just tried to use a `for` loop here, but it fails with an ICE:

```
error: internal compiler error: compiler/rustc_passes/src/check_const.rs:142:17: we should not have reached this point, since `.await` is denied earlier
```

joboet reviewed (commented):
I love the approach, encapsulating the invariant is a very good idea. One small issue left.

I changed the PR title to make it fit the contents, feel free to adapt it further.

reitermarkus committed:
Improve escape methods.

reitermarkus commented:
@rustbot ready

reitermarkus commented:
@joboet, I now reverted to using `Range<u8>` since that version generates fewer instructions, and also added a hint when converting `char` to `u32` whihch allows optimizing the index operations.

joboet commented on library/core/src/char/convert.rs:
I'm not *opposed* to these changes, but I feel deeply uncomfortable merging them as part of this PR. For one, these kind of changes tend to have quite some impact on performance, which should be evaluated separately from the other optimizations here. Also, I think it would make more sense to add the `assume` when lowering the `as` in the backend, so that every cast can profit from it – but I'm not the right reviewer for compiler stuff.

Could I therefore ask you to move these changes to a separate PR? We can then go ahead and merge this one.

reitermarkus replied:
Moved to separate PR.

reitermarkus commented:
@rustbot ready

joboet commented on library/core/src/char/mod.rs:
```diff
@@ -219,18 +218,24 @@ impl fmt::Display for EscapeUnicode {
-    fn backslash(chr: ascii::Char) -> Self {
-        let data = [ascii::Char::ReverseSolidus, chr];
-        Self(escape::EscapeIterInner::from_array(data))
+    #[inline]
+    const fn ascii(c: u8) -> Self {
+        Self(escape::EscapeIterInner::ascii(c))
     }
```
```suggestion
```

joboet commented on library/core/src/char/mod.rs:
```diff
@@ -219,18 +218,24 @@ impl fmt::Display for EscapeUnicode {
-        Self(escape::EscapeIterInner::from_array(data))
+    #[inline]
+    const fn printable(c: ascii::Char) -> Self {
+        Self::ascii(c.to_u8())
```
I think `ascii` is only used here? Then it doesn't make sense to define a separate function...
```suggestion
        Self(escape::EscapeIterInner::ascii(c.to_u8()))
```

joboet commented on library/core/src/ascii.rs:
```diff
@@ -91,17 +91,21 @@ pub struct EscapeDefault(escape::EscapeIterInner<4>);
     pub(crate) fn empty() -> Self {
-        let data = [Char::Null; 4];
-        EscapeDefault(escape::EscapeIterInner::new(data, 0..0))
+        EscapeDefault(escape::EscapeIterInner::empty())
```
I don't really mind either `EscapeDefault` or `Self`, but it should be consistent.
```suggestion
        Self(escape::EscapeIterInner::empty())
```

joboet reviewed (approved):
This looks great, I just have two small nits left.

joboet commented:
Cool stuff, thank you!
@bors r+

bors commented:
:pushpin: Commit 4edf12d33e0aa67a82e6587e94469f3761cb1f53 has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 4edf12d33e0aa67a82e6587e94469f3761cb1f53 -->
<!-- homu: {"type":"Approved","sha":"4edf12d33e0aa67a82e6587e94469f3761cb1f53","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 4edf12d33e0aa67a82e6587e94469f3761cb1f53 with merge ade234d5743795423db6cc7cd52541390a088eab...
<!-- homu: {"type":"BuildStarted","head_sha":"4edf12d33e0aa67a82e6587e94469f3761cb1f53","merge_sha":"ade234d5743795423db6cc7cd52541390a088eab"} -->

fmease commented:
Perf regression in https://github.com/rust-lang/rust/pull/125144#issuecomment-2113157992

reitermarkus commented:
`escape_ascii` produces 13 fewer instructions:

https://rust.godbolt.org/z/8eob57os8

`escape_unicode` produces the same number of instructions:

https://rust.godbolt.org/z/zv9f8eb1n
